system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -69,6 +70,92 @@ class TopSite extends React.Component {
   }
 }
 
+/**
+ * A proxy class that uses double requestAnimationFrame from
+ * componentDidMount to dispatch a SAVE_SESSION_PERF_DATA to the main procsess
+ * after the paint.
+ *
+ * This uses two callbacks because, after one callback, this part of the tree
+ * may have rendered but not yet reflowed.  This strategy is modeled after
+ * https://stackoverflow.com/a/34999925 but uses a double rFA because
+ * we want to get to the closest reliable paint for measuring, and
+ * setTimeout is often throttled or queued by browsers in ways that could
+ * make it lag too long.
+ *
+ * XXX Should probably be split out into a higher-order component to wrap
+ * whatever.
+ *
+ * @class TopSitesPerfTimer
+ * @extends {React.Component}
+ */
+class TopSitesPerfTimer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.perfSvc = this.props.perfSvc || perfSvc;
+
+    this._sendPaintedEvent = this._sendPaintedEvent.bind(this);
+    this._timestampSent = false;
+  }
+
+  componentDidMount() {
+    this._maybeSendPaintedEvent();
+  }
+
+  componentDidUpdate() {
+    this._maybeSendPaintedEvent();
+  }
+
+  /**
+   * Call the given callback when the subsequent animation frame
+   * (not the upcoming one) paints.
+   *
+   * @param {Function} callback
+   *
+   * @returns void
+   */
+  _onNextFrame(callback) {
+    requestAnimationFrame(() => {
+      requestAnimationFrame(callback);
+    });
+  }
+
+  _maybeSendPaintedEvent() {
+    // If we've already saved a timestamp for this session, don't do so again.
+    if (this._timestampSent) {
+      return;
+    }
+
+    // We don't want this to ever happen, but sometimes it does.  And when it
+    // does (typically on the first newtab at startup time calling
+    // componentDidMount), the paint(s) we care about will be later (eg
+    // in a subsequent componentDidUpdate).
+    if (!this.props.TopSites.initialized) {
+      // XXX should send bad event
+      return;
+    }
+
+    this._onNextFrame(this._sendPaintedEvent);
+  }
+
+  _sendPaintedEvent() {
+    // XXX should collapse these two calls below by adding a
+    // perfSvc.markAndGetAbsStart() method
+    this.perfSvc.mark("topsites_first_painted_ts");
+    let topsites_first_painted_ts = this.perfSvc
+      .getMostRecentAbsMarkStartByName("topsites_first_painted_ts");
+
+    this.props.dispatch(ac.SendToMain({
+      type: at.SAVE_SESSION_PERF_DATA,
+      data: {topsites_first_painted_ts}
+    }));
+
+    this._timestampSent = true;
+  }
+  render() {
+    return (<TopSites {...this.props} />);
```

Comment:
Personally I would probably do

```js
<PerfTimer>
   <TopSites />
</PerfTimer>
```

and render `this.props.children` instead of nesting but this works too.